File: electron/lib/extractor.js
// 发票文本提取与正则字段解析（Electron 桌面版，pdfjs 提取 + 正则）
// 与 Python 版 extractor.py 行为一致：兼容单栏/双栏版式、金额不依赖 ¥ 符号
const path = require('path');
const fs = require('fs');

// pdfjs 是 ESM 模块；Electron 内置 Node 不支持 require(esm)，用动态 import
let pdfjsPromise = null;
function getPdfjs() {
  if (!pdfjsPromise) pdfjsPromise = import('pdfjs-dist/legacy/build/pdf.mjs');
  return pdfjsPromise;
}

// pdfjs v4 的 Node 用法：工厂是实例（非类），主线程直接调 fetch(data)
// - cMap:  fetch({name}) → { cMapData: Uint8Array, isCompressed: true }
// - fonts: fetch({filename}) → Buffer（原始字体数据）
const PDFJS_PKG = path.dirname(require.resolve('pdfjs-dist/package.json'));
const CMAPS_DIR = path.join(PDFJS_PKG, 'cmaps');
const STD_FONTS_DIR = path.join(PDFJS_PKG, 'standard_fonts');

class NodeCMapReaderFactory {
  constructor() {}
  async fetch({ name }) {
    const data = fs.readFileSync(path.join(CMAPS_DIR, name + '.bcmap'));
    return { cMapData: new Uint8Array(data), isCompressed: true };
  }
}

class NodeStandardFontDataFactory {
  constructor() {}
  async fetch({ filename }) {
    return fs.readFileSync(path.join(STD_FONTS_DIR, filename));
  }
}

async function openDoc(data) {
  const pdfjs = await getPdfjs();
  return pdfjs.getDocument({
    data,
    CMapReaderFactory: NodeCMapReaderFactory,
    cMapPacked: true,
    StandardFontDataFactory: NodeStandardFontDataFactory,
    cMapUrl: null,
    standardFontDataUrl: null,
    useWorkerFetch: false,
    isEvalSupported: false,
  }).promise;
}

const FIELD_LABELS = {
  invoice_no: '发票号码',
  date: '开票日期',
  seller: '销售方',
  buyer: '购买方',
  amount_excl: '金额',
  tax: '税额',
  amount: '价税合计',
  amount_cn: '金额大写',
  type: '票种',
  seller_tax_id: '销售方税号',
  buyer_tax_id: '购买方税号',
};
const KEY_FIELDS = ['invoice_no', 'date', 'amount'];

const RE_INVOICE_NO_CTX = /发票号码[:：]?\s*([0-9]{8,25})/;
const RE_INVOICE_NO_ALONE = /(?<![0-9])(\d{20})(?!\d)/;
const RE_DATE_CTX = /开票日期[:：]?\s*(\d{4})[年\-/.](\d{1,2})[月\-/.](\d{1,2})/;
const RE_DATE = /(\d{4})[年\-/.](\d{1,2})[月\-/.](\d{1,2})/;
const RE_TAX_ID = /(?:纳税人识别号|统一社会信用代码)[^\dA-Z]*([0-9A-Z]{15,20})/;
const RE_AMOUNT_CN = /价税合计[（(]大写[）)]\s*([\u4e00-\u9fa5零壹贰叁肆伍陆柒捌玖拾佰仟万亿元角分整]+)/;
const RE_NAME_INLINE = /名称\s*[:：]\s*([^\s:：]{2,60})/;
const RE_COMPANY = /([\u4e00-\u9fa5A-Za-z0-9·（）()]{4,40}(?:公司|厂|店|事务所|中心|集团|医院|学校))/;

function emptyFields() {
  const f = {};
  for (const k of Object.keys(FIELD_LABELS)) f[k] = '';
  return f;
}

// 无状态 matchAll 包装（group=0 返回整段，否则返回捕获组）
function allMatches(re, s, group = 0) {
  const out = [];
  const g = new RegExp(re.source, re.flags.includes('g') ? re.flags : re.flags + 'g');
  let m;
  while ((m = g.exec(s)) !== null) out.push(m[group]);
  return out;
}

// ---------- 文本提取（含双栏） ----------
async function extractTexts(pdfPath) {
  const data = new Uint8Array(fs.readFileSync(pdfPath));
  const doc = await openDoc(data);
  const fullLines = [], leftLines = [], rightLines = [];
  for (let i = 1; i <= doc.numPages; i++) {
    const page = await doc.getPage(i);
    const width = page.getViewport({ scale: 1 }).width;
    const mid = width / 2;
    const tc = await page.getTextContent();
    const rows = groupIntoLines(tc.items);
    for (const line of rows) {
      fullLines.push(line.text);
      const left = line.items.filter(it => it.x < mid).map(it => it.str).join('');
      const right = line.items.filter(it => it.x >= mid).map(it => it.str).join('');
      if (left.trim()) leftLines.push(left);
      if (right.trim()) rightLines.push(right);
    }
  }
  await doc.destroy();
  return { full: fullLines.join('\n'), left: leftLines.join('\n'), right: rightLines.join('\n') };
}

// 按 y 分组成行（y 容差 3），行内按 x 排序，相邻间隙 >2 插入空格
function groupIntoLines(items) {
  const rows = [];
  const Y_TOL = 3;
  for (const it of items) {
    if (!it.str || !it.str.trim()) continue;
    const x = it.transform[4];
    const y = it.transform[5];
    let row = rows.find(r => Math.abs(r.y - y) <= Y_TOL);
    if (!row) {
      row = { y, items: [] };
      rows.push(row);
    }
    row.items.push({ x, x1: x + (it.width || 0), str: it.str });
  }
  rows.sort((a, b) => b.y - a.y); // PDF 坐标 y 向上，大 y = 页面上方
  return rows.map(row => {
    row.items.sort((a, b) => a.x - b.x);
    let text = '';
    let prevX1 = null;
    for (const it of row.items) {
      if (prevX1 !== null && it.x - prevX1 > 2) text += ' ';
      text += it.str;
      prevX1 = it.x1;
    }
    return { text, items: row.items };
  });
}

// ---------- 字段解析 ----------
function detectType(text) {
  const head = text.slice(0, 800);
  if (head.includes('铁路电子客票')) return '铁路电子客票';
  if (head.includes('通行费') && (head.includes('电子发票') || head.includes('数电票'))) return '通行费电子发票（普通）';
  if (head.includes('数电票') || head.includes('全面数字化电子发票')) return head.includes('专用') ? '数电票（专用）' : '数电票（普通）';
  if (head.includes('电子发票')) return head.includes('专用') ? '电子专用发票' : '电子普通发票';
  if (head.includes('增值税专用发票')) return '增值税专用发票';
  if (head.includes('增值税普通发票')) return '增值税普通发票';
  if (head.includes('专用发票')) return '专用发票';
  if (head.includes('普通发票')) return '普通发票';
  return '';
}

function blockAfter(text, keywords, nLines = 8) {
  const lines = text.split('\n').map(l => l.trim()).filter(Boolean);
  for (let i = 0; i < lines.length; i++) {
    if (keywords.some(k => lines[i].includes(k))) return lines.slice(i, i + nLines);
  }
  return [];
}

// 去除竖排拆字残留（"购买方信息/销售方信息"竖排拆字可能黏在名称前后：购销买售）
function cleanName(name) {
  return String(name || '')
    .replace(/^[购销买售:：\s]+/, '')
    .replace(/[统一社会信用代码纳税人识别号:：\s]+$/, '')
    .trim();
}

function partyFromColumn(colText) {
  const out = { name: '', tax_id: '' };
  const lines = colText.trim().replace(/名\s+称/g, '名称').split('\n').map(l => l.trim()).filter(Boolean);
  if (!lines.length) return out;
  for (let i = 0; i < lines.length; i++) {
    const m = RE_NAME_INLINE.exec(lines[i]);
    if (m) { out.name = cleanName(m[1]); if (out.name) break; }
    else if (/名称\s*[:：]?$/.test(lines[i]) && i + 1 < lines.length) {
      // 「名称:」单独一行，名称在下一行（pdfjs 对双栏版式的常见布局）
      out.name = cleanName(lines[i + 1]);
      if (out.name) break;
    }
  }
  if (!out.name) {
    const mm = RE_COMPANY.exec(colText.replace(/名\s+称/g, '名称'));
    if (mm) out.name = cleanName(mm[1]);
  }
  const tm = RE_TAX_ID.exec(colText);
  if (tm) out.tax_id = tm[1];
  return out;
}

function partyBlock(text, keyword) {
  const out = { name: '', tax_id: '' };
  const block = blockAfter(text, [keyword + '信息', keyword + '名称', keyword]);
  const joined = block.join('\n').replace(/名\s+称/g, '名称');
  let m = RE_NAME_INLINE.exec(joined);
  if (m) out.name = m[1].trim();
  else {
    for (const line of block) {
      if (/公司|厂|店|事务所|中心|集团|医院|学校/.test(line) && line.length <= 80) {
        out.name = line.trim();
        break;
      }
    }
  }
  m = RE_TAX_ID.exec(joined);
  if (m) out.tax_id = m[1];
  return out;
}

function extractAmount(text) {
  // 1) 价税合计行：优先「小写」后金额；否则带小数金额；否则段内最后数字
  const idx = text.indexOf('价税合计');
  if (idx !== -1) {
    const seg = text.slice(idx, idx + 120);
    let m = /小写[)）]?\s*[¥￥´]?\s*([\d,]+(?:\.\d{1,2})?)/.exec(seg);
    if (!m) {
      const dec = allMatches(/[\d,]+\.\d{1,2}/, seg);
      if (dec.length) m = { 1: dec[dec.length - 1] };
      else {
        const nums = allMatches(/[\d,]+(?:\.\d{1,2})?/, seg);
        if (nums.length) m = { 1: nums[nums.length - 1] };
      }
    }
    if (m) return (m[1] || '').replace(/,/g, '');
  }
  // 2) 合计行：金额+税额 求和
  let total = null;
  for (const line of text.split('\n')) {
    if (line.includes('合计') && !line.includes('价税合计')) {
      const vals = allMatches(/[\d,]+(?:\.\d{1,2})?/, line).map(s => parseFloat(s.replace(/,/g, '')));
      if (vals.length) total = (total || 0) + vals.reduce((a, b) => a + b, 0);
    }
  }
  if (total !== null) return total.toFixed(2);
  // 3) 兜底：最后 ¥/´ 金额
  const amounts = allMatches(/[¥￥´]\s*([\d,]+(?:\.\d{1,2})?)/, text, 1);
  if (amounts.length) return amounts[amounts.length - 1].replace(/,/g, '');
  return '';
}

// 金额（不含税）与税额：锚定「合计」行（非价税合计），行内两个带小数数字 = 金额、税额
// 例：`合 计 ¥293.81 ¥38.19` → amount_excl=293.81, tax=38.19
function extractAmountExclAndTax(text) {
  const out = { amount_excl: '', tax: '' };
  for (const line of text.split('\n')) {
    const l = line.replace(/\s/g, '');
    if (l.includes('合计') && !l.includes('价税合计')) {
      const nums = allMatches(/[\d,]+\.\d{1,2}/, l);
      if (nums.length >= 2) {
        out.amount_excl = nums[nums.length - 2].replace(/,/g, '');
        out.tax = nums[nums.length - 1].replace(/,/g, '');
        break;
      }
    }
  }
  return out;
}

// 折叠连续重复的短语（防复制水印 PDF：标签/短语重复 3 遍）
// '名称：名称：名称：' → '名称：'；'购购购' → '购'；'发票号码：发票号码：发票号码：' → '发票号码：'
// 只折叠 ≥3 次重复（\1{2,}），保护合法叠字（如公司名中恰好重复 2 次的情况）
function dedup(text) {
  return String(text).replace(/(.{1,12}?)\1{2,}/g, '$1');
}

// 水印残留的伪名称：只含「名称/购销买售」等标签字的匹配结果（如 '名称'、'名称名称'）
const JUNK_NAME = /^(?:名称|购销买售|名|称)+$/;

function parseFields(text, leftCol = '', rightCol = '') {
  // 第一遍：原始文本解析（原始优先，保护合法叠字公司名）
  const fields = parseFieldsCore(text, leftCol, rightCol);
  const missing = KEY_FIELDS.some(k => !fields[k]) || !fields.type;
  // 名称类字段可能解析出「名称」这类水印残留伪值（非空但有误）
  const nameBad = ['seller', 'buyer'].some(k => fields[k] && JUNK_NAME.test(fields[k]));
  if (missing || nameBad) {
    // 第二遍：折叠重复字后的文本再解析，只补空值/伪值
    const dFields = parseFieldsCore(dedup(text), dedup(leftCol), dedup(rightCol));
    for (const k of Object.keys(fields)) {
      const bad = (k === 'seller' || k === 'buyer') && fields[k] && JUNK_NAME.test(fields[k]);
      if ((!fields[k] || bad) && dFields[k]) fields[k] = dFields[k];
    }
  }
  return fields;
}

function parseFieldsCore(text, leftCol = '', rightCol = '') {
  const fields = emptyFields();
  fields.type = detectType(text);

  let m = RE_INVOICE_NO_CTX.exec(text);
  if (!m) m = RE_INVOICE_NO_ALONE.exec(text);
  if (m) fields.invoice_no = m[1];

  m = RE_DATE_CTX.exec(text);
  if (!m) m = RE_DATE.exec(text);
  if (m) {
    const [, y, mo, d] = m;
    // 汉字年月日格式：2026年08月15日（月日补零）
    fields.date = `${String(+y).padStart(4, '0')}年${String(+mo).padStart(2, '0')}月${String(+d).padStart(2, '0')}日`;
  }

  fields.amount = extractAmount(text);

  // 金额（不含税）/税额：优先合计行直接提取；缺失一侧时用 价税合计 推算
  const ae = extractAmountExclAndTax(text);
  fields.amount_excl = ae.amount_excl;
  fields.tax = ae.tax;
  const total = parseFloat(fields.amount);
  if (!isNaN(total)) {
    if (fields.amount_excl && !fields.tax) {
      fields.tax = Math.max(0, Math.round((total - parseFloat(fields.amount_excl)) * 100) / 100).toFixed(2);
    } else if (fields.tax && !fields.amount_excl) {
      fields.amount_excl = Math.max(0, Math.round((total - parseFloat(fields.tax)) * 100) / 100).toFixed(2);
    }
  }

  m = RE_AMOUNT_CN.exec(text);
  if (m) fields.amount_cn = m[1];

  let seller = partyFromColumn(rightCol);
  let buyer = partyFromColumn(leftCol);
  if (!seller.name) seller = partyBlock(text, '销售方');
  if (!buyer.name) buyer = partyBlock(text, '购买方');
  fields.seller = seller.name;
  fields.seller_tax_id = seller.tax_id;
  fields.buyer = buyer.name;
  fields.buyer_tax_id = buyer.tax_id;
  return fields;
}

async function parsePdf(pdfPath) {
  let texts;
  try {
    texts = await extractTexts(pdfPath);
  } catch (e) {
    return { fields: emptyFields(), errors: [`PDF 读取失败: ${e.message}`], rawText: '' };
  }
  if (!texts.full.trim()) {
    return { fields: emptyFields(), errors: ['未提取到文本（可能不是文字型 PDF）'], rawText: '' };
  }
  const fields = parseFields(texts.full, texts.left, texts.right);
  const errors = [];
  for (const k of KEY_FIELDS) {
    if (!fields[k]) errors.push(`未识别到${FIELD_LABELS[k]}`);
  }
  return { fields, errors, rawText: texts.full };
}

module.exports = {
  FIELD_LABELS, KEY_FIELDS, extractTexts, parseFields, parseFieldsCore, parsePdf, emptyFields,
};

File: extractor.py
"""发票文本提取与正则字段解析（中国电子发票 / 数电票，文字型 PDF）。

兼容两类版式：
- 单栏：购买方信息/销售方信息分行（数电票常见）
- 双栏：购/销双方信息左右分栏、名称同行（增值税专用发票常见）
"""
import re

import pdfplumber

FIELD_LABELS = {
    "invoice_no": "发票号码",
    "date": "开票日期",
    "seller": "销售方",
    "buyer": "购买方",
    "amount": "价税合计",
    "amount_cn": "金额大写",
    "type": "票种",
    "seller_tax_id": "销售方税号",
    "buyer_tax_id": "购买方税号",
}

# 关键字段：缺这些就算"识别不完整"
KEY_FIELDS = ("invoice_no", "date", "amount")

RE_INVOICE_NO_CTX = re.compile(r"发票号码[:：]?\s*([0-9]{8,25})")
RE_INVOICE_NO_ALONE = re.compile(r"(?<![0-9])(\d{20})(?!\d)")
RE_DATE = re.compile(r"(\d{4})[年\-/.](\d{1,2})[月\-/.](\d{1,2})")
RE_DATE_CTX = re.compile(r"开票日期[:：]?\s*(\d{4})[年\-/.](\d{1,2})[月\-/.](\d{1,2})")
RE_TAX_ID = re.compile(r"(?:纳税人识别号|统一社会信用代码)[^\dA-Z]*([0-9A-Z]{15,20})")
RE_AMOUNT_NUM = re.compile(r"[\d,]+(?:\.\d{1,2})?")
RE_AMOUNT_CN = re.compile(
    r"价税合计[（(]大写[）)]\s*([\u4e00-\u9fa5零壹贰叁肆伍陆柒捌玖拾佰仟万亿元角分整]+)"
)
RE_NAME_INLINE = re.compile(r"名称\s*[:：]\s*([^\s:：]{2,60})")
RE_COMPANY = re.compile(r"([\u4e00-\u9fa5A-Za-z0-9·（）()]{4,40}(?:公司|厂|店|事务所|中心|集团|医院|学校))")


def _norm_name(text: str) -> str:
    """把 '名 称' 这类拆字还原为 '名称'（真实发票竖排/字距导致）。"""
    return re.sub(r"名\s+称", "名称", text)


def _dedup(text: str) -> str:
    """折叠连续重复字符。部分税务 PDF 做防复制水印：每个字重复 3 遍
    （'发票号码' → '发发发票票票号号号码码码'），导致 3+ 字短语匹配失败。"""
    return re.sub(r"(.)\1+", r"\1", text)


def extract_texts(pdf_path: str):
    """返回 (full_text, left_col, right_col)。
    left/right 按页面中线裁剪，用于双栏版式（左=购买方，右=销售方）。
    """
    full, left, right = [], [], []
    with pdfplumber.open(str(pdf_path)) as pdf:
        for page in pdf.pages:
            w, h = float(page.width), float(page.height)
            t = page.extract_text()
            if t:
                full.append(t)
            try:
                mid = w / 2
                lt = page.crop((0, 0, mid, h)).extract_text()
                rt = page.crop((mid, 0, w, h)).extract_text()
            except Exception:
                lt = rt = ""
            if lt:
                left.append(lt)
            if rt:
                right.append(rt)
    return "\n".join(full), "\n".join(left), "\n".join(right)


def extract_text(pdf_path: str) -> str:
    """提取全文（单栏视图，兼容旧调用）。"""
    full, _, _ = extract_texts(pdf_path)
    return full


def _detect_type(text: str) -> str:
    head = text[:800]
    if "铁路电子客票" in head:
        return "铁路电子客票"
    if "通行费" in head and ("电子发票" in head or "数电票" in head):
        return "通行费电子发票（普通）"
    if "数电票" in head or "全面数字化电子发票" in head:
        return "数电票（专用）" if "专用" in head else "数电票（普通）"
    if "电子发票" in head:
        return "电子专用发票" if "专用" in head else "电子普通发票"
    if "增值税专用发票" in head:
        return "增值税专用发票"
    if "增值税普通发票" in head:
        return "增值税普通发票"
    if "专用发票" in head:
        return "专用发票"
    if "普通发票" in head:
        return "普通发票"
    return ""


def _block_after(text: str, keywords, n_lines: int = 8):
    """找到关键词所在行之后的若干行（含该行），用于单栏版式。"""
    lines = [l.strip() for l in text.splitlines() if l.strip()]
    for i, line in enumerate(lines):
        if any(k in line for k in keywords):
            return lines[i : i + n_lines]
    return []


def _party_from_column(col_text: str) -> dict:
    """双栏版式：单栏文本内提取 名称 + 税号。"""
    out = {"name": "", "tax_id": ""}
    c = _norm_name(col_text.strip())
    if not c:
        return out
    m = RE_NAME_INLINE.search(c)
    if m:
        out["name"] = m.group(1).strip()
    else:
        m = RE_COMPANY.search(c)
        if m:
            out["name"] = m.group(1).strip()
    m = RE_TAX_ID.search(c)
    if m:
        out["tax_id"] = m.group(1)
    return out


def _party_block(text: str, keyword: str) -> dict:
    """单栏版式：按信息块提取 名称 + 税号。"""
    out = {"name": "", "tax_id": ""}
    block = _block_after(text, [keyword + "信息", keyword + "名称", keyword])
    joined = _norm_name("\n".join(block))
    m = RE_NAME_INLINE.search(joined)
    if m:
        out["name"] = m.group(1).strip()
    else:
        for line in block:
            if re.search(r"公司|厂|店|事务所|中心|集团|医院|学校", line) and len(line) <= 80:
                out["name"] = line.strip()
                break
    m = RE_TAX_ID.search(joined)
    if m:
        out["tax_id"] = m.group(1)
    return out


def _extract_amount(text: str) -> str:
    """价税合计提取，不依赖 ¥ 符号（真实 PDF 中可能提取为 ´ 等）。"""
    # 1) 价税合计行：优先取「小写」后紧跟的金额；否则取段内带小数的货币金额（备注里的整数编号不干扰）
    idx = text.find("价税合计")
    if idx != -1:
        seg = text[idx : idx + 120]
        m = re.search(r"小写[)）]?\s*[¥￥´]?\s*([\d,]+(?:\.\d{1,2})?)", seg)
        if not m:
            dec = re.findall(r"[\d,]+\.\d{1,2}", seg)
            if dec:
                m = dec[-1]
            else:
                nums = RE_AMOUNT_NUM.findall(seg)
                m = nums[-1] if nums else None
        if m:
            return (m if isinstance(m, str) else m.group(1)).replace(",", "")
    # 2) 合计行：金额 + 税额 求和（增值税发票 价税合计 = 金额 + 税额）
    total = None
    for line in text.splitlines():
        if "合计" in line and "价税合计" not in line:
            vals = [float(n.replace(",", "")) for n in RE_AMOUNT_NUM.findall(line)]
            if vals:
                total = sum(vals)
    if total is not None:
        return f"{total:.2f}"
    # 3) 兜底：全文最后一个 ¥/´ 金额
    amounts = re.findall(r"[¥￥´]\s*([\d,]+(?:\.\d{1,2})?)", text)
    if amounts:
        return amounts[-1].replace(",", "")
    return ""


def parse_fields(text: str, left_col: str = "", right_col: str = "") -> dict:
    """正则解析，返回字段字典。

    先按原始文本解析；缺失字段用「连续重复字折叠」后的文本再解析一次
    （兼容防复制水印 PDF）。原始结果优先，避免误伤合法叠字（如公司名）。
    """
    fields = _parse_fields_core(text, left_col, right_col)
    if missing_key_fields(fields) or not fields.get("type"):
        d_text = _dedup(text)
        d_fields = _parse_fields_core(d_text, _dedup(left_col), _dedup(right_col))
        for k in fields:
            if not fields[k] and d_fields[k]:
                fields[k] = d_fields[k]
    return fields


def _parse_fields_core(text: str, left_col: str = "", right_col: str = "") -> dict:
    """单遍正则解析（供 parse_fields 调用）。"""
    fields = {k: "" for k in FIELD_LABELS}
    fields["type"] = _detect_type(text)

    m = RE_INVOICE_NO_CTX.search(text)
    if not m:
        m = RE_INVOICE_NO_ALONE.search(text)
    if m:
        fields["invoice_no"] = m.group(1)

    m = RE_DATE_CTX.search(text)
    if not m:
        m = RE_DATE.search(text)
    if m:
        y, mo, d = m.groups()
        fields["date"] = f"{int(y):04d}-{int(mo):02d}-{int(d):02d}"

    fields["amount"] = _extract_amount(text)

    m = RE_AMOUNT_CN.search(text)
    if m:
        fields["amount_cn"] = m.group(1)

    # 双栏版式：右栏=销售方，左栏=购买方；单栏版式：信息块兜底
    seller = _party_from_column(right_col)
    buyer = _party_from_column(left_col)
    if not seller["name"]:
        seller = _party_block(text, "销售方")
    if not buyer["name"]:
        buyer = _party_block(text, "购买方")
    fields["seller"] = seller["name"]
    fields["seller_tax_id"] = seller["tax_id"]
    fields["buyer"] = buyer["name"]
    fields["buyer_tax_id"] = buyer["tax_id"]
    return fields


def missing_key_fields(fields: dict):
    return [k for k in KEY_FIELDS if not fields.get(k)]


def parse_pdf(pdf_path: str) -> dict:
    """解析单个 PDF，返回 {fields, errors, text}。"""
    errors = []
    try:
        full, left, right = extract_texts(pdf_path)
    except Exception as e:
        return {"fields": {k: "" for k in FIELD_LABELS},
                "errors": [f"PDF 读取失败: {e}"], "text": ""}
    if not full.strip():
        errors.append("未提取到文本（可能不是文字型 PDF）")
    fields = parse_fields(full, left, right) if full.strip() else {k: "" for k in FIELD_LABELS}
    for k in missing_key_fields(fields):
        errors.append(f"未识别到{FIELD_LABELS[k]}")
    return {"fields": fields, "errors": errors, "text": full}
